src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PreReleaseUserService.cs

```diff
@@ -231,130 +234,185 @@ public class PreReleaseUserService : IPreReleaseUserService
+                    if (failure != null)
+                    {
+                        return failure;
+                    }
 
-            await _context.SaveChangesAsync();
-            return Unit.Instance;
+                    return Unit.Instance;
+                });
         }
 
-        private async Task<Either<ActionResult, Unit>> ValidateNewPreReleaseUser(Guid releaseId, string email)
+        private async Task<Either<ActionResult, PreReleaseUserViewModel>> InvitePreReleaseUser(Release release,
+            string email)
         {
-            var hasExistingReleaseRole = await _context
-                .UserReleaseRoles
-                .Include(r => r.User)
-                .Where(
-                    r =>
-                        r.ReleaseId == releaseId
-                        && r.User.Email.ToLower() == email.ToLower()
-                        && r.Role == ReleaseRole.PrereleaseViewer
-                )
-                .AnyAsync();
-
-            if (hasExistingReleaseRole)
+            var user = await _userRepository.FindByEmail(email);
+
+            if (user == null)
             {
-                return ValidationActionResult(UserAlreadyExists);
+                await CreateUserInvite(email);
+                return await CreateUserReleaseInvite(release, email)
+                    .OnSuccess(_ => new PreReleaseUserViewModel(email));
             }
 
-            var hasExistingReleaseInvite = await _context
-                .UserReleaseInvites
-                .AsQueryable()
-                .Where(
-                    i =>
-                        i.ReleaseId == releaseId
-                        && i.Email.ToLower() == email.ToLower()
-                         && i.Role == ReleaseRole.PrereleaseViewer
-                )
-                .AnyAsync();
-
-            if (hasExistingReleaseInvite)
+            return await CreateAcceptedUserReleaseInvite(release, email, user)
+                .OnSuccess(_ => new PreReleaseUserViewModel(email));
+        }
+
+        private async Task<Either<ActionResult, Unit>> CreateUserReleaseInvite(Release release, string email)
+        {
+            if (!await UserHasPreReleaseInvite(release.Id, email))
             {
-                return ValidationActionResult(UserAlreadyExists);
+                var sendEmail = release.ApprovalStatus == ReleaseApprovalStatus.Approved;
+                await _context.AddAsync(
+                    new UserReleaseInvite
+                    {
+                        Email = email.ToLower(),
+                        ReleaseId = release.Id,
+                        Role = ReleaseRole.PrereleaseViewer,
+                        EmailSent = sendEmail,
+                        Created = UtcNow,
+                        CreatedById = _userService.GetUserId()
+                    }
+                );
+                await _context.SaveChangesAsync();
```

Comment:
Thanks, I've made this so that it never gets saved if the email is not sent. I also reordered things so that the UserInvite in the `UsersAndRolesDbContext` is not created either if this fails and added a unit test for this.